### PR TITLE
Datum var cleanup

### PR DESCRIPTION
**Pull request by Cyberboss**
Datums should always ALWAYS be as lightweight as possible. Credit to @vuonojenmustaturska for the idea

Moved `ui_screen` to `/datum/tgui`
Moved `fingerprintslast` to `/atom` (Only possible thanks to forensics components. Well done @kevinz000)
Removed `focusers` if something doesn't fix it's focus by the time it's qdeleted, it's an error @ninjanomnom

### CODE REVIEW

**Comment by coiax** [comment]
If the dummy subtype is so problematic, why do we even have it?

File: code/datums/datum.dm [hunk 1]
```diff
@@ -1,14 +1,13 @@
 /datum
-    var/gc_destroyed //Time when this object was destroyed.
-    var/list/active_timers  //for SStimer
-    var/list/datum_components //for /datum/components
-    var/ui_screen = "home"  //for tgui
```
Comment (JJRcop) [inline code/datums/datum.dm]: Quite embarrassing that this managed to get through

File: code/datums/datumvars.dm [hunk 2]
```diff
@@ -1,18 +1,11 @@
...
 	return TRUE
 
 /datum/proc/vv_edit_var(var_name, var_value) //called whenever a var is edited
-	switch(var_name)
-		if ("vars")
-			return FALSE
-		if ("var_edited")
-			return FALSE
-	var_edited = TRUE
+	if(var_name == NAMEOF(src, vars) || (var_name == NAMEOF(src, datum_flags) && (!isnum(var_value) || (!(var_value & DF_VAR_EDITED) && (datum_flags & DF_VAR_EDITED)))))
```
Comment (nicbn) [inline code/datums/datumvars.dm]: There's no problem on letting datum_flags be altered, as line 8 will add the var edited flag later.

**Review by nicbn — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
@nicbn it's mainly to preserve the old behaviour

On Jan 12, 2018 9:19 PM, "nicbn" <notifications@github.com> wrote:

*@nicbn* requested changes on this pull request.
------------------------------

In code/datums/datumvars.dm
<https://github.com/tgstation/tgstation/pull/34314#discussion_r161360750>:

>  /datum/proc/can_vv_get(var_name)
 	return TRUE

 /datum/proc/vv_edit_var(var_name, var_value) //called whenever a var is edited
-	switch(var_name)
-		if ("vars")
-			return FALSE
-		if ("var_edited")
-			return FALSE
-	var_edited = TRUE
+	if(var_name == NAMEOF(src, vars) || (var_name == NAMEOF(src,
datum_flags) && (!isnum(var_value) || (!(var_value & DF_VAR_EDITED) &&
(datum_flags & DF_VAR_EDITED)))))

There's no problem on letting datum_flags be altered, as line 8 will add
the var edited flag later.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/34314#pullrequestreview-88645759>,
or mute the thread

**Comment by nicbn** [comment]
The behaviour will be the same, the only thing that will change is the performance, which will increase.

File: code/datums/datum.dm [hunk 3]
```diff
@@ -1,14 +1,13 @@
 /datum
-    var/gc_destroyed //Time when this object was destroyed.
-    var/list/active_timers  //for SStimer
-    var/list/datum_components //for /datum/components
-    var/ui_screen = "home"  //for tgui
-    var/use_tag = FALSE
-    var/datum/weakref/weak_reference
+	var/gc_destroyed //Time when this object! was destroyed.
```
Comment (nicbn) [inline code/datums/datum.dm]: Why change this comment?

File: code/datums/datumvars.dm (hunk 2 above)
Comment (AnturK) [reply code/datums/datumvars.dm]: You could change it to some invalid value but i agree this is probably unneeded and will just complicate things in the future.

**Pull request merged**